### PR TITLE
Add Terraform templates

### LINKED ISSUE #316: TF/CloudFormation template for deploying EG (with standalone etcd cluster)
Would be great to have TF or CF template/config for showing how to deploy example setup of EG with a standalone etcd cluster.

### PR DESCRIPTION
**Pull request by baniol**
## What did you implement:

Closes #316

## Todos:

- [ ] Write tests
- [x] Write documentation
- [ ] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [ ] Provide verification commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Terraform templates for Event Gateway
```
Comment (mthenw) [inline terraform/README.md]: 1. Can you move all those file to `/contrib/terraform` folder.
2. Ideally would be to expose it as a module that someone else can use. So I think this readme should be mainly about how to use this module.

File: terraform/network.tf [hunk 2]
```diff
@@ -0,0 +1,60 @@
+data "aws_availability_zones" "available" {}
```
Comment (mthenw) [inline terraform/network.tf]: What do you think about using existing module for creating VPC and all network stuff https://registry.terraform.io/modules/terraform-aws-modules/vpc/aws/1.37.0?

File: terraform/alb.tf [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  egress {
+    from_port   = 0
+    to_port     = 0
+    protocol    = "-1"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+}
+
+# Config load balancer
+resource "aws_lb" "config" {
```
Comment (mthenw) [inline terraform/alb.tf]: This module should probably simplify the setup here https://registry.terraform.io/modules/terraform-aws-modules/alb/aws/3.4.0?tab=resources

File: terraform/etcd.tf [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module "etcd" {
```
Comment (mthenw) [inline terraform/etcd.tf]: What do you think about using existing module from CoreOS https://github.com/coreos/tectonic-installer/tree/track-1/modules/aws/etcd?

File: terraform/README.md (hunk 1 above)
Comment (baniol) [reply terraform/README.md]: Not a problem. I'll refactor it in that direction

File: terraform/network.tf (hunk 2 above)
Comment (baniol) [reply terraform/network.tf]: You're right. Using some community modules makes sense.

File: terraform/alb.tf (hunk 3 above)
Comment (baniol) [reply terraform/alb.tf]: The problem with this module is that it doesn't seem to allow to control target group health checks. We need to overwrite the default "/" path with "/v1/status". What's more, the events API server doesn't expose any health check endpoint returning 200 code (the root returns 202).

Comment (mthenw) [reply terraform/alb.tf]: Good point but looks like it should be possible.

screenshot 2018-07-06 09 10 09

Comment (baniol) [reply terraform/alb.tf]: yes, I've missed the default section. Works perfectly with the alb module.

File: terraform/etcd.tf (hunk 4 above)
Comment (baniol) [reply terraform/etcd.tf]: This one will be harder to use as it is. Etcd is integrated with the k8s installer and has some internal dependencies, like dns, certs or s3 bucket with init scripts.
I need to take a deeper look.

File: contrib/example.tf [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module "event-gateway" {
+  source = "terraform"
```
Comment (mthenw) [inline contrib/example.tf]: Could you change it to full github URL so people can just copy/paste?

Comment (baniol) [reply contrib/example.tf]: sure, I'll do a little refactor and add a readme file when I'm done with the etcd module

File: .gitignore [hunk 6]
```diff
@@ -8,3 +8,4 @@ default.etcd*
 coverage.txt
 profile.out
 tests/testing.etcd*
+.terraform
```
Comment (mthenw) [inline .gitignore]: we don't need this here if it's only a module

File: contrib/example.tf (hunk 5 above)
Comment (mthenw) [reply contrib/example.tf]: awesome thanks!

**Comment by baniol** [comment]
@mthenw A few notes regarding the latest commit:

* the coreos etcd module is a part of Kubernetes setup in Tectonic installer and has some dependencies, like certificates, coreos ignition, dns. I've extracted those dependencies into the etcd wrapper (`modules/etcd`)
* Tectonic Installer repo is under Apache v2 licence, so I've included the NOTICE file
* it's possible to use https for the etcd (`tls_enabled` param), however eg libkv store does not initialize with tls config. I don't think this it a big issue for now, as the etcd cluster runs in private subnets.

**Comment by mthenw** [comment]
@baniol overall looks good. We need few days to review it. Again, thanks for contribution.

**Pull request merged**